File: androidApp/src/main/kotlin/zinqq/app/screens/settings/PeersScreen.kt
package zinqq.app.screens.settings

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.KeyboardCapitalization
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextDecoration
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.draw.drawBehind
import kotlinx.coroutines.launch
import uniffi.wallet_core.ChannelStateLabel
import uniffi.wallet_core.ChannelView
import uniffi.wallet_core.PeerView
import zinqq.app.theme.ZinqqTheme

/**
 * The PWA's Peers (U17, R10 UI; `Peers.tsx`): the connect input — parsed
 * client-side like the PWA, then handed to OpenChannel where the core's
 * `open_channel` connects if needed — the `(N connected, M saved)` list with
 * Refresh, per-peer status dots and channel-guarded Forget, and the nested
 * channel rows with Close / Force Close actions.
 */
@Composable
fun PeersScreen(
    port: SettingsPort,
    onBack: (() -> Unit)?,
    onOpenChannel: (String) -> Unit,
    onCloseChannel: (channelId: String, force: Boolean) -> Unit,
) {
    val colors = ZinqqTheme.colors
    val scope = rememberCoroutineScope()

    var peerAddress by remember { mutableStateOf("") }
    var connectError by remember { mutableStateOf<String?>(null) }
    var peers by remember { mutableStateOf<List<PeerView>?>(null) }
    var channels by remember { mutableStateOf<Map<String, List<ChannelView>>>(emptyMap()) }
    var loadError by remember { mutableStateOf<String?>(null) }
    var forgetError by remember { mutableStateOf<String?>(null) }

    fun refresh() {
        scope.launch {
            try {
                val fetchedPeers = port.listPeers()
                channels = channelsByPeer(port.listChannels())
                peers = fetchedPeers
                loadError = null
            } catch (e: Exception) {
                loadError = e.message ?: "Lightning node error"
            }
        }
    }

    LaunchedEffect(Unit) { refresh() }

    fun connect() {
        connectError = null
        when (val parsed = parsePeerAddress(peerAddress.trim())) {
            is PeerAddressParse.Invalid -> connectError = parsed.message
            is PeerAddressParse.Valid -> onOpenChannel(peerAddress.trim())
        }
    }

    SettingsScaffold(title = "Peers", onBack = onBack) {
        val currentPeers = peers
        if (currentPeers == null && loadError == null) {
            CenteredNote("Loading Lightning node...")
            return@SettingsScaffold
        }
        if (loadError != null && currentPeers == null) {
            Column(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(horizontal = 24.dp)
                    .padding(top = 64.dp),
                horizontalAlignment = Alignment.CenterHorizontally,
            ) {
                Text(
                    text = "Lightning node error",
                    color = colors.onDark,
                    fontWeight = FontWeight.SemiBold,
                    fontSize = 16.sp,
                )
                Text(
                    text = loadError.orEmpty(),
                    color = colors.danger,
                    fontSize = 14.sp,
                    textAlign = TextAlign.Center,
                    modifier = Modifier.padding(top = 8.dp),
                )
            }
            return@SettingsScaffold
        }

        Column(
            modifier = Modifier
                .fillMaxWidth()
                .verticalScroll(rememberScrollState())
                .padding(horizontal = 24.dp)
                .padding(top = 8.dp, bottom = 32.dp),
            verticalArrangement = Arrangement.spacedBy(20.dp),
        ) {
            // Connect form (Peers.tsx:167-193).
            Column {
                Text(
                    text = "Connect & Open Channel",
                    color = colors.onDarkMuted,
                    fontSize = 14.sp,
                    fontWeight = FontWeight.Medium,
                )
                Box(
                    modifier = Modifier
                        .padding(top = 8.dp)
                        .fillMaxWidth()
                        .clip(RoundedCornerShape(12.dp))
                        .background(colors.darkElevated)
                        .padding(horizontal = 16.dp, vertical = 14.dp),
                ) {
                    BasicTextField(
                        value = peerAddress,
                        onValueChange = { peerAddress = it },
                        singleLine = true,
                        keyboardOptions = KeyboardOptions(
                            capitalization = KeyboardCapitalization.None,
                            autoCorrectEnabled = false,
                        ),
                        textStyle = TextStyle(
                            color = colors.onDark,
                            fontFamily = FontFamily.Monospace,
                            fontSize = 14.sp,
                        ),
                        cursorBrush = SolidColor(colors.hot),
                        modifier = Modifier.fillMaxWidth(),
                        decorationBox = { innerTextField ->
                            if (peerAddress.isEmpty()) {
                                Text(
                                    text = "pubkey@host:port",
                                    color = colors.onDarkMuted,
                                    fontFamily = FontFamily.Monospace,
                                    fontSize = 14.sp,
                                )
                            }
                            innerTextField()
                        },
                    )
                }
                connectError?.let {
                    Text(
                        text = it,
                        color = colors.danger,
                        fontSize = 14.sp,
                        modifier = Modifier.padding(top = 8.dp),
                    )
                }
                SettingsCta(
                    label = "Next",
                    background = colors.cta,
                    contentColor = colors.onCta,
                    enabled = peerAddress.isNotBlank(),
                    disabledAlpha = 0.3f,
                    onClick = ::connect,
                    modifier = Modifier.padding(top = 8.dp),
                )
            }

            // Peer list (Peers.tsx:196-310).
            Column {
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    Text(
                        text = peersCountLabel(currentPeers.orEmpty()),
                        color = colors.onDarkMuted,
                        fontSize = 14.sp,
                        fontWeight = FontWeight.Medium,
                        modifier = Modifier.weight(1f),
                    )
                    Text(
                        text = "Refresh",
                        color = colors.onDark,
                        fontSize = 12.sp,
                        textDecoration = TextDecoration.Underline,
                        modifier = Modifier
                            .clip(RoundedCornerShape(6.dp))
                            .clickable { refresh() }
                            .padding(8.dp),
                    )
                }
                forgetError?.let {
                    Text(
                        text = it,
                        color = colors.danger,
                        fontSize = 14.sp,
                        modifier = Modifier.padding(top = 4.dp),
                    )
                }
                if (currentPeers.isNullOrEmpty()) {
                    Text(
                        text = "No peers connected",
                        color = colors.onDarkMuted,
                        fontSize = 14.sp,
                        textAlign = TextAlign.Center,
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(vertical = 16.dp),
                    )
                } else {
                    Column(
                        modifier = Modifier.padding(top = 8.dp),
                        verticalArrangement = Arrangement.spacedBy(8.dp),
                    ) {
                        currentPeers.forEach { peer ->
                            PeerCard(
                                peer = peer,
                                channels = channels[peer.pubkey].orEmpty(),
                                onForget = {
                                    forgetError = null
                                    scope.launch {
                                        try {
                                            port.forgetPeer(peer.pubkey)
                                            refresh()
                                        } catch (e: Exception) {
                                            forgetError = forgetErrorMessage(e)
                                        }
                                    }
                                },
                                onCloseChannel = onCloseChannel,
                            )
                        }
                    }
                }
            }
        }
    }
}

@Composable
private fun PeerCard(
    peer: PeerView,
    channels: List<ChannelView>,
    onForget: () -> Unit,
    onCloseChannel: (channelId: String, force: Boolean) -> Unit,
) {
    val colors = ZinqqTheme.colors
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(12.dp))
            .background(colors.darkElevated)
            .padding(16.dp),
        verticalArrangement = Arrangement.spacedBy(8.dp),
    ) {
        Row(verticalAlignment = Alignment.CenterVertically) {
            Box(
                modifier = Modifier
                    .size(10.dp)
                    .clip(CircleShape)
                    .background(if (peer.connected) colors.success else colors.onDarkMuted),
            )
            Text(
                text = peerDisplayId(peer.pubkey),
                color = colors.onDark,
                fontFamily = FontFamily.Monospace,
                fontSize = 13.sp,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
                modifier = Modifier
                    .weight(1f)
                    .padding(horizontal = 12.dp),
            )
            Text(
                text = peerStatusLabel(peer.connected),
                color = if (peer.connected) colors.success else colors.onDarkMuted,
                fontSize = 12.sp,
                fontWeight = FontWeight.SemiBold,
            )
            if (showsForget(peer)) {
                val enabled = forgetEnabled(peer)
                Text(
                    text = "Forget",
                    color = colors.danger,
                    fontSize = 12.sp,
                    modifier = Modifier
                        .padding(start = 12.dp)
                        .alpha(if (enabled) 1f else 0.3f)
                        .clip(RoundedCornerShape(6.dp))
                        .clickable(enabled = enabled, onClick = onForget)
                        .padding(4.dp),
                )
            }
        }
        channels.forEach { channel ->
            ChannelRow(channel = channel, onCloseChannel = onCloseChannel)
        }
    }
}

@Composable
private fun ChannelRow(
    channel: ChannelView,
    onCloseChannel: (channelId: String, force: Boolean) -> Unit,
) {
    val colors = ZinqqTheme.colors
    val closing = channel.state == ChannelStateLabel.CLOSING
    // `ml-5 border-l pl-3` (Peers.tsx:253): indented with a left border rule.
    Column(
        modifier = Modifier
            .padding(start = 20.dp)
            .fillMaxWidth()
            .leftBorder(colors.darkBorder)
            .padding(start = 12.dp),
        verticalArrangement = Arrangement.spacedBy(4.dp),
    ) {
        Row(
            modifier = Modifier.fillMaxWidth(),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Text(
                text = channelStateText(channel.state),
                color = if (closing) colors.warning else colors.onDarkMuted,
                fontSize = 12.sp,
                fontWeight = if (closing) FontWeight.SemiBold else FontWeight.Normal,
                modifier = Modifier.weight(1f),
            )
            Text(
                text = channelCapacityText(channel),
                color = colors.onDark,
                fontSize = 12.sp,
                fontWeight = FontWeight.SemiBold,
            )
        }
        if (closing) {
            Text(
                text = CLOSING_IN_PROGRESS_NOTE,
                color = colors.onDarkMuted,
                fontSize = 12.sp,
            )
        }
        Row(
            modifier = Modifier.fillMaxWidth(),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Row(
                modifier = Modifier.weight(1f),
                horizontalArrangement = Arrangement.spacedBy(12.dp),
            ) {
                Text(
                    text = channelSendText(channel),
                    color = colors.onDarkMuted,
                    fontSize = 12.sp,
                )
                Text(
                    text = channelReceiveText(channel),
                    color = colors.onDarkMuted,
                    fontSize = 12.sp,
                )
                channelReserveText(channel)?.let {
                    Text(text = it, color = colors.onDarkMuted, fontSize = 12.sp)
                }
            }
            Text(
                text = channelCloseActionLabel(channel.state),
                color = colors.danger,
                fontSize = 12.sp,
                fontWeight = FontWeight.SemiBold,
                modifier = Modifier
                    .clip(RoundedCornerShape(6.dp))
                    .clickable { onCloseChannel(channel.channelId, closing) }
                    .padding(4.dp),
            )
        }
    }
}

/** A 1dp left border rule, like the PWA's `border-l border-dark-border`. */
private fun Modifier.leftBorder(color: Color): Modifier = drawBehind {
    drawLine(
        color = color,
        start = androidx.compose.ui.geometry.Offset(0f, 0f),
        end = androidx.compose.ui.geometry.Offset(0f, size.height),
        strokeWidth = 1.dp.toPx(),
    )
}

File: iosApp/iosApp/Screens/Settings/PeersScreen.swift
import Shared
import SwiftUI

/// The PWA's Peers (U22, R10 UI; `Peers.tsx`): the connect input — parsed
/// client-side like the PWA, then handed to OpenChannel where the core's
/// `openChannel` connects if needed — the `(N connected, M saved)` list with
/// Refresh, per-peer status dots and channel-guarded Forget, and the nested
/// channel rows with Close / Force Close actions. Mirrors Android's
/// `PeersScreen`.
struct PeersScreen: View {
    let port: any SettingsPort
    var onBack: (() -> Void)?
    let onOpenChannel: (String) -> Void
    let onCloseChannel: (_ channelId: String, _ force: Bool) -> Void

    @Environment(\.zinqqColors) private var colors
    @State private var peerAddress = ""
    @State private var connectError: String?
    @State private var peers: [PeerView]?
    @State private var channels: [String: [ChannelView]] = [:]
    @State private var loadError: String?
    @State private var forgetError: String?

    var body: some View {
        SettingsScaffold(title: "Peers", onBack: onBack) {
            if peers == nil && loadError == nil {
                CenteredSettingsNote("Loading Lightning node...")
            } else if let loadError, peers == nil {
                loadErrorBody(loadError)
            } else {
                content
            }
        }
        .task { await refresh() }
    }

    private func refresh() async {
        do {
            let fetchedPeers = try await port.listPeers()
            channels = channelsByPeer(try await port.listChannels())
            peers = fetchedPeers
            loadError = nil
        } catch {
            loadError = kotlinThrowable(error)?.message
                ?? (error as NSError).localizedDescription
        }
    }

    private func connect() {
        connectError = nil
        let trimmed = peerAddress.trimmingCharacters(in: .whitespacesAndNewlines)
        switch parsePeerAddress(trimmed) {
        case let .invalid(message):
            connectError = message
        case .valid:
            onOpenChannel(trimmed)
        }
    }

    private func loadErrorBody(_ message: String) -> some View {
        VStack(spacing: 8) {
            Text("Lightning node error")
                .font(ZinqqFont.sans(16, weight: .semibold))
                .foregroundColor(colors.onDark)
            Text(message)
                .font(ZinqqFont.sans(14))
                .foregroundColor(colors.danger)
                .multilineTextAlignment(.center)
        }
        .padding(.horizontal, 24)
        .padding(.top, 64)
        .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .top)
    }

    private var content: some View {
        ScrollView {
            VStack(spacing: 20) {
                connectForm
                peerList
            }
            .padding(.horizontal, 24)
            .padding(.top, 8)
            .padding(.bottom, 32)
        }
    }

    /// Connect form (Peers.tsx:167-193).
    private var connectForm: some View {
        VStack(alignment: .leading, spacing: 8) {
            Text("Connect & Open Channel")
                .font(ZinqqFont.sans(14, weight: .medium))
                .foregroundColor(colors.onDarkMuted)
            TextField(
                "",
                text: $peerAddress,
                prompt: Text("pubkey@host:port")
                    .font(.system(size: 14, design: .monospaced))
                    .foregroundColor(colors.onDarkMuted)
            )
            .font(.system(size: 14, design: .monospaced))
            .foregroundColor(colors.onDark)
            .tint(colors.hot)
            .textInputAutocapitalization(.never)
            .autocorrectionDisabled(true)
            .keyboardType(.asciiCapable)
            .padding(.horizontal, 16)
            .padding(.vertical, 14)
            .background(colors.darkElevated)
            .clipShape(RoundedRectangle(cornerRadius: 12))
            .accessibilityLabel("Peer address")
            if let connectError {
                Text(connectError)
                    .font(ZinqqFont.sans(14))
                    .foregroundColor(colors.danger)
            }
            SettingsCta(
                label: "Next",
                background: colors.cta,
                contentColor: colors.onCta,
                action: connect,
                enabled: !peerAddress.trimmingCharacters(in: .whitespaces).isEmpty,
                disabledAlpha: 0.3
            )
        }
    }

    /// Peer list (Peers.tsx:196-310).
    private var peerList: some View {
        VStack(alignment: .leading, spacing: 8) {
            HStack {
                Text(peersCountLabel(peers ?? []))
                    .font(ZinqqFont.sans(14, weight: .medium))
                    .foregroundColor(colors.onDarkMuted)
                Spacer()
                Button {
                    Task { await refresh() }
                } label: {
                    Text("Refresh")
                        .font(ZinqqFont.sans(12))
                        .underline()
                        .foregroundColor(colors.onDark)
                        .padding(8)
                }
                .accessibilityLabel("Refresh")
            }
            if let forgetError {
                Text(forgetError)
                    .font(ZinqqFont.sans(14))
                    .foregroundColor(colors.danger)
            }
            if let peers, !peers.isEmpty {
                VStack(spacing: 8) {
                    ForEach(peers, id: \.pubkey) { peer in
                        PeerCard(
                            peer: peer,
                            channels: channels[peer.pubkey] ?? [],
                            onForget: {
                                forgetError = nil
                                Task {
                                    do {
                                        try await port.forgetPeer(pubkey: peer.pubkey)
                                        await refresh()
                                    } catch {
                                        forgetError = forgetErrorMessage(error)
                                    }
                                }
                            },
                            onCloseChannel: onCloseChannel
                        )
                    }
                }
            } else {
                Text("No peers connected")
                    .font(ZinqqFont.sans(14))
                    .foregroundColor(colors.onDarkMuted)
                    .frame(maxWidth: .infinity)
                    .padding(.vertical, 16)
            }
        }
    }
}

private struct PeerCard: View {
    let peer: PeerView
    let channels: [ChannelView]
    let onForget: () -> Void
    let onCloseChannel: (_ channelId: String, _ force: Bool) -> Void

    @Environment(\.zinqqColors) private var colors

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            HStack(spacing: 12) {
                Circle()
                    .fill(peer.connected ? colors.success : colors.onDarkMuted)
                    .frame(width: 10, height: 10)
                Text(peerDisplayId(peer.pubkey))
                    .font(.system(size: 13, design: .monospaced))
                    .foregroundColor(colors.onDark)
                    .lineLimit(1)
                    .truncationMode(.tail)
                Spacer()
                Text(peerStatusLabel(connected: peer.connected))
                    .font(ZinqqFont.sans(12, weight: .semibold))
                    .foregroundColor(peer.connected ? colors.success : colors.onDarkMuted)
                if showsForget(peer) {
                    let enabled = forgetEnabled(peer)
                    Button(action: onForget) {
                        Text("Forget")
                            .font(ZinqqFont.sans(12))
                            .foregroundColor(colors.danger)
                            .padding(4)
                            .opacity(enabled ? 1 : 0.3)
                    }
                    .disabled(!enabled)
                    .accessibilityLabel("Forget peer")
                }
            }
            ForEach(channels, id: \.channelId) { channel in
                ChannelRow(channel: channel, onCloseChannel: onCloseChannel)
            }
        }
        .padding(16)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(colors.darkElevated)
        .clipShape(RoundedRectangle(cornerRadius: 12))
    }
}

private struct ChannelRow: View {
    let channel: ChannelView
    let onCloseChannel: (_ channelId: String, _ force: Bool) -> Void

    @Environment(\.zinqqColors) private var colors

    var body: some View {
        let closing = channel.state == .closing
        // `ml-5 border-l pl-3` (Peers.tsx:253): indented with a left rule.
        VStack(alignment: .leading, spacing: 4) {
            HStack {
                Text(channelStateText(channel.state))
                    .font(ZinqqFont.sans(12, weight: closing ? .semibold : .regular))
                    .foregroundColor(closing ? colors.warning : colors.onDarkMuted)
                Spacer()
                Text(channelCapacityText(channel))
                    .font(ZinqqFont.sans(12, weight: .semibold))
                    .foregroundColor(colors.onDark)
            }
            if closing {
                Text(closingInProgressNote)
                    .font(ZinqqFont.sans(12))
                    .foregroundColor(colors.onDarkMuted)
            }
            HStack(spacing: 12) {
                Text(channelSendText(channel))
                    .font(ZinqqFont.sans(12))
                    .foregroundColor(colors.onDarkMuted)
                Text(channelReceiveText(channel))
                    .font(ZinqqFont.sans(12))
                    .foregroundColor(colors.onDarkMuted)
                if let reserve = channelReserveText(channel) {
                    Text(reserve)
                        .font(ZinqqFont.sans(12))
                        .foregroundColor(colors.onDarkMuted)
                }
                Spacer()
                Button {
                    onCloseChannel(channel.channelId, closing)
                } label: {
                    Text(channelCloseActionLabel(channel.state))
                        .font(ZinqqFont.sans(12, weight: .semibold))
                        .foregroundColor(colors.danger)
                        .padding(4)
                }
                .accessibilityLabel(channelCloseActionLabel(channel.state))
            }
        }
        .padding(.leading, 12)
        .overlay(alignment: .leading) {
            Rectangle()
                .fill(colors.darkBorder)
                .frame(width: 1)
        }
        .padding(.leading, 20)
    }
}
